core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +70,157 @@ class SYCLInternal {
 
   std::unique_ptr<cl::sycl::queue> m_queue;
 
+  // USMObjectMem is a reusable buffer for a single object
+  // in USM memory
+  template <sycl::usm::alloc Kind>
+  class USMObjectMem {
+   public:
+    static constexpr sycl::usm::alloc kind = Kind;
+
+    friend void swap(USMObjectMem& lhs, USMObjectMem& rhs) noexcept {
+      using std::swap;
+      swap(lhs.m_q, rhs.m_q);
+      swap(lhs.m_data, rhs.m_data);
+      swap(lhs.m_capacity, rhs.m_capacity);
+    }
+
+    USMObjectMem()                    = default;
+    USMObjectMem(USMObjectMem const&) = delete;
+    USMObjectMem& operator=(USMObjectMem const&) = delete;
+
+    USMObjectMem(USMObjectMem&& that) noexcept
+        : m_q(std::move(that.m_q)),
+          m_data(that.m_data),
+          m_capacity(that.m_capacity) {
+      that.m_data     = nullptr;
+      that.m_capacity = 0;
+    }
+
+    USMObjectMem& operator=(USMObjectMem&& that) noexcept {
+      swap(*this, that);
+      return *this;
+    }
+
+    ~USMObjectMem() { sycl::free(m_data, m_q); }
+
+    explicit USMObjectMem(sycl::queue q) noexcept : m_q(std::move(q)) {}
+
+    sycl::queue queue() const noexcept { return m_q; }
+
+    void* data() noexcept { return m_data; }
+    const void* data() const noexcept { return m_data; }
+
+    size_t capacity() const noexcept { return m_capacity; }
+
+    // reserve() allocates space for at least n bytes
+    // returns the new capacity
+    size_t reserve(size_t n) {
+      if (m_capacity < n) {
+        // First free what we have (in case malloc can reuse it)
+        sycl::free(m_data, m_q);
+
+        m_data = sycl::malloc(n, m_q, kind);
+        if (!m_data) {
+          m_capacity = 0;
+          throw std::bad_alloc();
+        }
+
+        m_capacity = n;
+      }
+
+      return m_capacity;
+    }
+
+    // This will memcpy an object T into memory held by this object
+    // returns: a T* to that object
+    //
+    // Note:  it is UB to dereference this pointer with an object that is
+    // not an implicit-lifetime nor trivially-copyable type, but presumably much
+    // faster because we can use USM device memory
+    template <typename T>
+    T* memcpy_from(const T& t) {
+      reserve(sizeof(T));
+      sycl::event memcopied = m_q.memcpy(m_data, std::addressof(t), sizeof(T));
+      memcopied.wait();
+
+      return reinterpret_cast<T*>(m_data);
```

Comment:
It is a raw pointer because it has no lifetime to manage.  It could be a unique_ptr with the same deleter though, since calling the pseudo-dtor has no effect.

Will make private.

Reply:
I was mainly concerned that it is annoying if the return type of `copy_from` depends on the function that is called internally and we have to find slightly annoying workarounds to accommodate for both a raw pointer and a `std::unique_ptr`. So I would return a `std::uniqu_ptr<T>` here. I am not feeling strongly about it, though.

Reply:
That annoyed me too.

Having given this more thought, I think it is less weird to have a `unique_ptr` which doesn't call the destructor for objects in USM device memory because they are trivially copyable, as well as always tracking whether or not the object in USMObjectMem is live.  I'll make that change.